### PR TITLE
ENH: motion v3, ppm/wfs pmps coordination

### PR DESCRIPTION
**Pull request by ZLLentz**
## Description

- Upgrade to the newest libraries and implement the imager/WFS coordination requirement
- Run pre-commit
- Switch to library version for AT1K4 control

## Motivation and Context

- Requirement: PF1K0 in = lower reqs on IM2K0
- New motion states enables the above and is good

## How Has This Been Tested?

Will test during PAMM

## Where Has This Been Documented?

Release notes I guess

## Pre-merge checklist
- [ ] Code works interactively
- [x] Code contains descriptive comments
- [ ] Test suite passes locally
- [x] Libraries are set to fixed versions and not ``Always Newest``
- [x] Code committed with ``pre-commit`` (alternatively ``pre-commit run --all-files``)

### CODE REVIEW

**Comment by ZLLentz** [comment]
I forgot to ask for reviews on this

I plan to test during the PAMM tomorrow and update as needed

**Comment by tongju12** [comment]
LGTM. Like the ST_PositionState, makes the code neat. TMO should modify like this later.

File: plc-kfe-motion/kfe_motion/GVLs/Main.TcGVL [hunk 1]
```diff
@@ -148,54 +148,54 @@ VAR_GLOBAL
...
     {attribute 'pytmc' := 'pv: IM1K4:XTES:CLZ'}
     {attribute 'TcLinkTo' := '.bHome                := TIIB[IM1K4-EL1088-E5]^Channel 2^Input;
                               .nRawEncoderUINT     := TIIB[IM1K4-EL7342-E4]^ENC Status compact Channel 1^Counter value'}
-    M31: DUT_MotionStage := (sName := 'IM1K4:XTES:CLZ');
+    M31: ST_MotionStage := (sName := 'IM1K4:XTES:CLZ');
     {attribute 'pytmc' := 'pv: IM1K4:XTES:CLF'}
     {attribute 'TcLinkTo' := '.bHome := TIIB[IM1K4-EL1088-E5]^Channel 4^Input;
-                              .nRawEncoderUINT     := TIIB[IM1K4-EL5042-E3]^ENC Status compact Channel 2^Counter value'}
-    M32: DUT_MotionStage := (sName := 'IM1K4:XTES:CLF');
+                              .nRawEncoderUINT     := TIIB[IM1K4-EL7342-E4]^ENC Status compact Channel 2^Counter value'}
```
Comment (klauer) [inline plc-kfe-motion/kfe_motion/GVLs/Main.TcGVL:189]: (Aside: EL7342 is new to me so I half-wondered if this was in error)

File: plc-kfe-motion/kfe_motion/POUs/PRG_AT1K4_SOLID.TcPOU [hunk 2]
```diff
@@ -99,344 +110,130 @@ END_IF
...
-fbStage1.stOut.bValid                      := TRUE;
-fbStage1.stOut.bMoveOk                     := TRUE;
-
-fbStage1.stFilter1.sName                   := '(1) 25 um C';
-fbStage1.stFilter1.fPosition               := 40.5;
-fbStage1.stFilter1.fDelta                  := DEFAULT_DELTA;
-fbStage1.stFilter1.fVelocity               := DEFAULT_VELOCITY;
-fbStage1.stFilter1.bValid                  := TRUE;
-fbStage1.stFilter1.bMoveOk                 := TRUE;
+fbStateSetup(stPositionState:=fbStage1.stOut, sName:='Out', fPosition:=22.70);
```
Comment (klauer) [inline plc-kfe-motion/kfe_motion/POUs/PRG_AT1K4_SOLID.TcPOU:120]: This is just so much cleaner 😭 💯

**Review by klauer — APPROVED** [review]
I tried to check for copy/paste errors here and couldn't spot any.
The changes look good to me 👍

File: plc-kfe-motion/kfe_motion/GVLs/Main.TcGVL (hunk 1 above)
Comment (ZLLentz) [reply plc-kfe-motion/kfe_motion/GVLs/Main.TcGVL:189]: This is just a weird one-off sort of terminal for the XTES imagers (DC lens motors)

File: plc-kfe-motion/kfe_motion/POUs/PRG_AT1K4_SOLID.TcPOU (hunk 2 above)
Comment (ZLLentz) [reply plc-kfe-motion/kfe_motion/POUs/PRG_AT1K4_SOLID.TcPOU:120]: I just spent like 15 minutes self-reviewing these to make sure I didn't copy/paste the positions- glad we both agree there wasn't a copy/paste error, excited to not need to look at the huge wall of set-one-attr again, glad that our previous brainstorming work in the motion library made such a nice result. Feels good

Comment (ZLLentz) [reply plc-kfe-motion/kfe_motion/POUs/PRG_AT1K4_SOLID.TcPOU:120]: I also liked how it easy it was for these to just have 2 different preset defaults- one for "good states" and one for "bad states" (without filters)

Comment (klauer) [reply plc-kfe-motion/kfe_motion/POUs/PRG_AT1K4_SOLID.TcPOU:120]: I didn't comment on that aspect but muttered to myself as I was reading it about how nice the 2-default-fb approach ended up.
Glad this worked out so well!  🥳

**Comment by ZLLentz** [comment]
Problems found in live:
- the core feature (cross-pmps between wfs and imager) doesn't work at all, my CASE statement always falls into the ELSE branch (???)
- at startup everything is in a raw transition state instead of the current position state

I'm going to go through and verify that everything still moves via state device move and then retreat to a comfier place to continue debug

**Comment by ZLLentz** [comment]
I don't think I've dropped any PVs but IM2K0's python class is starting to throw a fit on startup that also breaks the screen:
```
raise ValueError('Invalid enum member name: {0}'.format(
ValueError: Invalid enum member name:
```
Looks like a blank string as an enum name maybe

**Comment by ZLLentz** [comment]
Actually I figured it out, I think everything is resolved except for the "transition state at startup" issue

**Comment by ZLLentz** [comment]
Transition state at startup appears to be a motion library bug- I'll test out a fix and link to that PR with the intention of merging/tagging that and using it in a tag here

**Comment by ZLLentz** [comment]
https://github.com/pcdshub/lcls-twincat-motion/pull/201
is only a problem on the first cycle, but that's enough to bug out the startup

**Comment by ZLLentz** [comment]
It seems like there is a further race condition that will require some offline thinking (not under PAMM pressure) to resolve.
It seems to be related to other intermittent possible cases where we either cannot determine the current position on the first cycle or we determine the wrong starting position.
In all of these cases, we're still safe- the only issue is that we come up in a state with weird PMPS. I _think_ this is related the order and timing of the PLC process startup vs the NC process but cannot prove it.
The weird state can be resolved after startup by either moving each motor or by resetting each FB with the clear errors button.

For now:
- tag motion library
- use motion library tag here
- merge this pr
- tag repo
- deploy tagged ioc and switch to it
- reset all the states devices
- other checks e.g. how are the cams doing

Later I can revisit the larger race condition world

**Comment by ZLLentz** [comment]
I'm setting the max states/motor counts based on the diagnostics from the motion library:
image

**Pull request merged**